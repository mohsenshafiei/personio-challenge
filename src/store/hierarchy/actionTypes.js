export default {
  INIT_APP: 'INIT_APP',
  INIT_APP_OK: 'INIT_APP_OK',
  JSON_FILE_UPLOADED: 'JSON_FILE_UPLOADED',
  JSON_FILE_UPLOADED_SUCCESSFUL: 'JSON_FILE_UPLOADED_SUCCESSFUL',
  REMOVE_PERSON: 'REMOVE_PERSON',
  ADD_PERSON: 'ADD_PERSON',
  CHANGE_FILTER: 'CHANGE_FILTER',
  TOGGLE_COLLAPSE: 'TOGGLE_COLLAPSE',
  DETECT_MULTIPLE_BOSS: 'DETECT_MULTIPLE_BOSS',
  REMOVE_MULTIPLE_PERSON: 'REMOVE_MULTIPLE_PERSON',
  NOTIFICATION: 'NOTIFICATION',
  CHANGE_HIERARCHY: 'CHANGE_HIERARCHY',
};
