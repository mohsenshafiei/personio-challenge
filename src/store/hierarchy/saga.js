import {
  put, takeEvery, select,
} from 'redux-saga/effects';

const getPerson = (employees = [], personId) => {
  let employee = employees.find(person => person.id === personId);
  if (employee) {
    return employee;
  }
  for (let i = 0; i < employees.length; i += 1) {
    employee = getPerson(employees[i].employees, personId);
    if (employee) {
      return employee;
    }
  }
  return undefined;
};

function* hierarchyChanged(action) {
  if (action.leaderId.indexOf(action.personId) === 0) {
    yield put({ type: 'NOTIFICATION', title: 'This Action Is Not Possible, You Are Making a Loop!', style: 'error' });
  } else {
    const person = yield select(state => getPerson(state.hierarchy.employees, action.personId));
    if (person) {
      yield put({ type: 'REMOVE_PERSON', personId: action.personId });
      yield put({ type: 'ADD_PERSON', leaderId: action.leaderId, person });
    }
  }
}
function* removeMultiplePerson(action) {
  yield put({ type: 'REMOVE_PERSON', personId: action.personId });
  yield put({ type: 'DETECT_MULTIPLE_BOSS' });
}

function* fileUpload() {
  // you can send it to server here
  yield put({ type: 'JSON_FILE_UPLOADED_SUCCESSFUL', payload: true });
}

export default [
  takeEvery('JSON_FILE_UPLOADED', fileUpload),
  takeEvery('CHANGE_HIERARCHY', hierarchyChanged),
  takeEvery('REMOVE_MULTIPLE_PERSON', removeMultiplePerson),
];
