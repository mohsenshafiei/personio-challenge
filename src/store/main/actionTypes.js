export default {
  CHANGE_LANGUAGE: 'CHANGE_LANGUAGE',
  NOTIFICATION_CALL: 'NOTIFICATION_CALL',
  JSON_FILE_UPLOADED_SUCCESSFUL: 'JSON_FILE_UPLOADED_SUCCESSFUL',
  SET_LANGUAGE: 'SET_LANGUAGE',
};
