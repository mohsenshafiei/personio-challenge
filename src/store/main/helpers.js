import i18n from '../../i18n';

export const delay = ms => new Promise(res => setTimeout(res, ms));

export const setLanguage = () => {
  if (i18n.language !== 'En') {
    i18n.lang = 'En';
    i18n.changeLanguage('En', (err) => {
      if (err) {
        return false;
      }
      window.localStorage.setItem('locale', 'En');
      return true;
    });
  } else {
    i18n.lang = 'De';
    i18n.changeLanguage('De', (err) => {
      if (err) {
        return false;
      }
      window.localStorage.setItem('locale', 'De');
      return true;
    });
  }
};
