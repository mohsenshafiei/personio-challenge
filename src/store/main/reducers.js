const initialState = {
  language: 'En',
};

const mainReducers = (state = initialState, action) => {
  switch (action.type) {
    case 'CHANGE_LANGUAGE': {
      return {
        ...state,
        language: action.payload,
      };
    }
    default:
      return state;
  }
};

export default mainReducers;
