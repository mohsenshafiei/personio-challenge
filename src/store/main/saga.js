import { takeEvery } from 'redux-saga/effects';
import i18n from '../../i18n';

function changeLanguage() {
  if (i18n.language !== 'en') {
    i18n.lang = 'en';
    i18n.changeLanguage('en', (err) => {
      if (err) {
        return false;
      }
      window.localStorage.setItem('locale', 'en');
      return true;
    });
  } else {
    i18n.lang = 'de';
    i18n.changeLanguage('de', (err) => {
      if (err) {
        return false;
      }
      window.localStorage.setItem('locale', 'de');
      return true;
    });
  }
}

export default [
  takeEvery('CHANGE_LANGUAGE', changeLanguage),
];
