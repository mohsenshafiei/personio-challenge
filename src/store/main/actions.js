import actionTypes from './actionTypes';

const changeLanguage = payload => (
  {
    type: actionTypes.SET_LANGUAGE,
    payload,
  }
);

export default changeLanguage;
