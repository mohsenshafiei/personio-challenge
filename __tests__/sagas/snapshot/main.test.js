import React from 'react';
import { changeLanguage, notification, fileUpload } from '../../../src/store/main/saga';
import actionTypes from "../../../src/store/main/actionTypes";
import { expectSaga } from 'redux-saga-test-plan';

describe('CHANGE LANGUAGE', () => {
  const action1 = { type: actionTypes.SET_LANGUAGE , payload: 'En' };
  it('change language snapshot', () => {
    return expectSaga(changeLanguage, action1)
      .run()
      .then((result) => {
        expect(result.toJSON()).toMatchSnapshot();
      });
  });
});

describe('Notification', () => {
  const action1 = { type: actionTypes.NOTIFICATION, style: 'error', title: 'This Action Is Not Possible, You Are Making a Loop!' };
  it('notification snapshot', () => {
    return expectSaga(notification, action1)
      .run()
      .then((result) => {
        expect(result.toJSON()).toMatchSnapshot();
      });
  });
});


describe('File Uploaded', () => {
  const action1 = { type: actionTypes.SAVE_JSON_FILE};
  it('file upload snapshot', () => {
    return expectSaga(fileUpload, action1)
      .run()
      .then((result) => {
        expect(result.toJSON()).toMatchSnapshot();
      });
  });
});
