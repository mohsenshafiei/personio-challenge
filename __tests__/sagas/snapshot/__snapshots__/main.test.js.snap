// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CHANGE LANGUAGE change language snapshot 1`] = `
Object {
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "payload": "De",
          "type": "CHANGE_LANGUAGE",
        },
        "channel": null,
      },
    },
  ],
}
`;

exports[`File Uploaded file upload snapshot 1`] = `
Object {
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "payload": true,
          "type": "JSON_FILE_UPLOADED_SUCCESSFUL",
        },
        "channel": null,
      },
    },
  ],
}
`;

exports[`Notification notification snapshot 1`] = `
Object {
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "payload": true,
          "style": "error",
          "title": "This Action Is Not Possible, You Are Making a Loop!",
          "type": "NOTIFICATION_CALL",
        },
        "channel": null,
      },
    },
  ],
}
`;
